src/mds/Locker.cc

```diff
@@ -225,6 +225,35 @@ struct MarkEventOnDestruct {
   }
 };
 
+/*
+ * Try to detect the possible deadlock between requests, and
+ * will try to break the newer request.
+ */
+void Locker::detect_dead_locks(const MDRequestRef& mdr, int lock_type)
+{
+  bool any_waiter = false;
+  bool order_mess = false;
+
+  dout(15) << "detect_dead_locks " << *mdr << dendl;
+  for (auto it = mdr->locks.begin(); it != mdr->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+
+    if (lock->is_waiter_for(SimpleLock::WAIT_ALL)) {
+      any_waiter = true;
+    }
+    if (lock->get_type() > lock_type) {
+      order_mess = true;
+    }
+    dout(15) << " lock " << *lock << " any_waiter " << any_waiter << " order_mess " << order_mess << dendl;
+  }
+
+  if (order_mess && any_waiter) {
```

Comment:
We should add another high-level warning (i.e. dout (1 < x < 5)) for `order_mess` alone, maybe even in health to make the corresponding FS runs fail and have our attention

Reply:
Sure, will fix it.